### PR TITLE
Add when directive

### PR DESCRIPTION
**Pull request by LarsDenBakker**
This PR adds a when directive, which can be used to efficiently switch between templates. 

Based on the condition, either the 'if' or the 'else' value is rendered. When switching between conditions, the previously rendered DOM is cached and re-used instead of recreated.

Caching is done by reparenting the the rendered content into a detached dom node when the condition switches. I didn't change any code in the actual library itself, but I think might be useful to create some extra helper functions for moving around part contents. Repeat could also make use of this.

I think this concept could be extended further by creating a 'switch' directive which switches between dom content and caches nodes.

### CODE REVIEW

**Comment by LarsDenBakker** [comment]
Yeah, I agree. Lazy evaluation is something we will need in other places, it'd be nice to handle this in the core somehow. Currently functions are rendered as a string, maybe lit could just unpack the function when passed as a value. Would that introduce much overhead for regular rendering?

**Comment by ruphin** [comment]
I think lazy evaluation is only necessary in a context where multiple results are possible.

If lit is to assume that all functions return results, should it always just call the function and render the results instead? What if the function returns another function? There's a lot of corner cases to handle here, and I don't think it should be the responsibility of core to do this.

The issue can easily be solved in the relevant directives, in this case:

```javascript
if (condition) {
  value is ifValue();
} else {
  parts.reverse();
  value = elseValue();
}
```

Which changes the API to:

```javascript
html`${when(condition, () => html`<div></div>`, () => html`<span></span>`)}`
```

It's just a question of what API you want the directive to have. As I mentioned above, I think any directive with multiple options should always be lazy-evaluated by default, because it is easy for developers to make mistakes otherwise.

File: src/lib/when.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+ * Creates cache consisting of a NodePart for each condition and a detached
+ * container used for caching rendered nodes.
+ *
+ * @param parentPart the parent part
+ */
+function createPartCache(parentPart: NodePart) {
+  const parts: PartCache =  [
+    new NodePart(parentPart.templateFactory),
+    new NodePart(parentPart.templateFactory),
+    document.createElement('div'),
```
[the remark this answers is not in the document]
Comment (LarsDenBakker) [reply src/lib/when.ts]: I've never worked with DocumentFragment before, I will look into that.

[the remark this answers is not in the document]
Comment (ruphin) [reply src/lib/when.ts]: I'm pretty sure you are familiar with it :)
For example, HTMLTemplateElement uses DocumentFragment to store the template contents:

https://developer.mozilla.org/en-US/docs/Web/API/HTMLTemplateElement/content

[the remark this answers is not in the document]
Comment (LarsDenBakker) [reply src/lib/when.ts]: Yeah, I've seen it before but just never used myself or looked much into it. Definitely is the better option here.

File: src/lib/when.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {directive, reparentNodes, DirectiveFn, NodePart} from '../core.js';
+
+type PartCache = [NodePart, NodePart, DocumentFragment];
```
Comment (ruphin) [inline src/lib/when.ts]: Is there a reason to have this an Array instead of an Object with named keys?

```javascript
{
  ifPart: NodePart,
  elsePart: NodePart,
  container: DocumentFragment
}
```

File: src/lib/when.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+ * `
+ *
+ * @param condition the condition to test against
+ * @param ifValue the content to render if
+ * @param elseValue
+ */
+export const when = (condition: boolean, ifValue: () => any, elseValue: () => any): DirectiveFn<NodePart> =>
+    directive((part: NodePart) => {
+      // Create cache if this was the first render
+      if (!partCaches.has(part)) {
```
Comment (justinfagnani) [inline src/lib/when.ts]: Using `Map#has()` causes there to be two map lookups in the hit case and three in the miss case. Use `Map#get()` instead and check for undefined:

```ts
let partCache = partCaches.get(part);
if (partCache === undefined) {
  partCache = {
    truePart: new NodePart(parentPart.templateFactory),
    falsePart: new NodePart(parentPart.templateFactory),
    container: document.createDocumentFragment(),
  };
  partCaches.set(part, partCache);
}
```

This is one lookup for a hit, two for a miss.

File: src/lib/when.ts (hunk 2 above)
Comment (justinfagnani) [reply src/lib/when.ts]: Agreed, but a small naming nit: since this isn't exactly an `if` statement, I'd name the properties semantically as `truePart` and `falsePart`

File: src/lib/when.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+      const partCache = partCaches.get(part)!;
+      const cacheContainer = partCache[2];
+      const parts = [partCache[0], partCache[1]];
+      let value;
+
+      if (condition) {
+        value = ifValue();
+      } else {
+        parts.reverse();
```
Comment (justinfagnani) [inline src/lib/when.ts]: this `reverse()` is a bit clever and hard to follow, IMO. Rather than a `parts` array, can you create two named variables?

File: src/lib/when.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+      if (!partCaches.has(part)) {
+        createPartCache(part);
+      }
+
+      const partCache = partCaches.get(part)!;
+      const cacheContainer = partCache[2];
+      const parts = [partCache[0], partCache[1]];
+      let value;
+
+      if (condition) {
```
Comment (justinfagnani) [inline src/lib/when.ts]: Before even checking the condition, I think we want to check that it's different from last time. If it's the same we can just propagate the current value.

File: src/lib/when.ts [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+      }
+
+      // If the old part was rendered, move it to the cache
+      if (prevPart.value) {
+        reparentNodes(cacheContainer, prevPart.startNode, prevPart.endNode.nextSibling);
+      }
+
+      // Set the new part's value
+      newPart.setValue(value);
+      newPart.commit();
```
Comment (justinfagnani) [inline src/lib/when.ts]: Shouldn't need a commit here.

File: src/test/lib/when_test.ts [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+      renderWhen(false);
+      assert.equal(stripExpressionDelimeters(container.innerHTML), '<span></span>');
+
+      renderWhen(true);
+      assert.equal(stripExpressionDelimeters(container.innerHTML), '<div></div>');
+    });
+
+    test('caches if/then templates', () => {
+      let ifEl;
```
Comment (justinfagnani) [inline src/test/lib/when_test.ts]: these can be const

Comment (LarsDenBakker) [reply src/test/lib/when_test.ts]: These are assigned values later on, so they can't be a const.

File: src/lib/when.ts (hunk 2 above)
Comment (LarsDenBakker) [reply src/lib/when.ts]: I see what you mean. I started out with just two parts, so I made it a tuple to keep it simple and because I did the reverse() later on. Later on I added the container element.

File: src/lib/when.ts (hunk 5 above)
Comment (LarsDenBakker) [reply src/lib/when.ts]: The value to set depends on the condition, even if it didn't change. So we still need to do some work. You're right though currently we do a lot of dom manipulation when not necessary.

I've separated the work for moving dom nodes when switching conditions.

File: src/lib/when.ts [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+export const when = (condition: boolean, trueValue: () => any, falseValue: () => any): DirectiveFn<NodePart> =>
+    directive((part: NodePart) => {
+      const partCache = getPartCache(part);
+      const { truePart, falsePart, cacheContainer, prevCondition } = partCache;
+
+      const nextPart = condition ? truePart : falsePart;
+      const nextValue = condition ? trueValue() : falseValue();
+
+      // Swap nodes if condition changed from previous
+      if (condition !== prevCondition) {
```
Comment (ruphin) [inline src/lib/when.ts]: You are checking for equality against the `prevCondition`, but what you really want is to check if they are both truthy or both falsy. Does TypeScript automatically add a boolean conversion to `condition` because it is typed as a boolean? If not, it might be worth to manually cast `condition` to a boolean value with `condition = !!condition` to make sure this check returns what you expect.

Comment (LarsDenBakker) [reply src/lib/when.ts]: Typescript forces consumers to use a boolean value, which I like. But for JS I guess the expectation that it checks truthiness instead. If we want to support that, then it needs some extra checks as now I depend on boolean equality so that the initial value of undefined for the cached condition does not cause it to match the false condition.

Comment (justinfagnani) [reply src/lib/when.ts]: Don't assume that consumer will be using TypeScript. You can rely on the type as much as you want to assert that the docs have to be followed exactly. In this case, JS users are so used to "truthiness" that we should support it. 

`condition = !!condition` is a fine way to do this, and you could even type `condition` as `unknown`. Make sure you test initial render in both the true and false cases.

File: src/lib/when.ts [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+/**
+ * @param parentPart the parent part
+ * @returns the cache for the given parent. creates a new cache if none exists
+ */
+function getPartCache(parentPart: NodePart) {
+  let cache = partCaches.get(parentPart);
+
+  // Create a new cache if this is the first render
+  if (!cache) {
```
Comment (justinfagnani) [inline src/lib/when.ts]: Use an explicit check instead of relying on boolean conversion: `if (cache === undefined)`

File: src/lib/when.ts [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+      nextPart.commit();
+
+      partCache.prevCondition = condition;
+    });
+
+/**
+ * @param parentPart the parent part
+ * @returns the cache for the given parent. creates a new cache if none exists
+ */
+function getPartCache(parentPart: NodePart) {
```
Comment (justinfagnani) [inline src/lib/when.ts]: Since this function is only called once I'd just linine it. Especially because it has side-effects on `parentPart`

File: src/lib/when.ts [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+ * @param condition the condition to test against
+ * @param trueValue the value to render given a true condition
+ * @param falseValue the value to render given a false condition
+ */
+export const when = (condition: boolean, trueValue: () => any, falseValue: () => any): DirectiveFn<NodePart> =>
+    directive((part: NodePart) => {
+      const partCache = getPartCache(part);
+      const { truePart, falsePart, cacheContainer, prevCondition } = partCache;
+
+      const nextPart = condition ? truePart : falsePart;
```
Comment (justinfagnani) [inline src/lib/when.ts]: Add some comments on what next and previous mean here.

**Comment by aadamsx** [comment]
> Don't assume that consumer will be using TypeScript. You can rely on the type as much as you want to assert that the docs have to be followed exactly. In this case, JS users are so used to "truthiness" that we should support it.

condition = !!condition is a fine way to do this, and you could even type condition as unknown. Make sure you test initial render in both the true and false cases.

I for one will be using this lib when it his 1.0 and WILL NOT be using Typescript to save my life. :)

**Comment by justinfagnani** [comment]
@LarsDenBakker #436 is submitted, so please rebase to master now.

**Comment by LarsDenBakker** [comment]
I rebased onto master
